File: doodle.js
// doodle.js
#!/usr/bin/env node
'use strict';

const blessed = require('blessed');
const fs = require('fs');
const path = require('path');
const os = require('os');

const RECORD_FILE = path.join(os.homedir(), '.doodle_record.json');

function loadRecord() {
    try { return JSON.parse(fs.readFileSync(RECORD_FILE)).record || 0; } catch { return 0; }
}
function saveRecord(record) {
    fs.writeFileSync(RECORD_FILE, JSON.stringify({ record }));
}

let speed = 50;
if (process.argv.includes('-s') && process.argv.length > process.argv.indexOf('-s')+1) {
    speed = parseInt(process.argv[process.argv.indexOf('-s')+1]) || 50;
}

const screen = blessed.screen({
    smartCSR: true,
    title: 'Doodle Jump',
    fullUnicode: true,
});

const height = screen.height;
const width = screen.width;
if (height < 20 || width < 30) {
    console.log('Terminal too small');
    process.exit(1);
}

let playerY = height - 3;
let playerX = Math.floor(width/2);
let velX = 0, velY = 0;
const gravity = 0.3, jump = -6;
let platforms = [];
let score = 0;
let best = loadRecord();
let gameOver = false;
let frameTime = speed;

for (let i=0; i<8; i++) {
    let x = Math.floor(Math.random() * (width-6)) + 1;
    let y = height - 2 - i*4;
    platforms.push({x, y, type: 0});
}

function draw() {
    screen.clear();
    // Player
    screen.fillRegion('@', playerX, playerY, playerX+1, playerY+1, blessed.colors.yellow, blessed.colors.black);
    // Platforms
    for (const p of platforms) {
        let color = blessed.colors.green;
        if (p.type === 1) color = blessed.colors.green; // moving (bold)
        else if (p.type === 2) color = blessed.colors.red;
        else if (p.type === 3) color = blessed.colors.red;
        for (let i=0; i<5; i++) {
            if (p.x+i >= 0 && p.x+i < width) {
                screen.fillRegion('#', p.x+i, p.y, p.x+i+1, p.y+1, color, blessed.colors.black);
            }
        }
    }
    // Score
    screen.setContent(0, 2, `Score: ${score}`, blessed.colors.cyan);
    screen.setContent(0, Math.floor(width/2)-4, `Best: ${best}`, blessed.colors.cyan);
    if (gameOver) {
        screen.setContent(Math.floor(height/2)-2, Math.floor(width/2)-15, `💀 Game Over! Score: ${score}  Best: ${best}`, blessed.colors.red);
        screen.setContent(Math.floor(height/2), Math.floor(width/2)-10, 'R - restart | Q - quit', blessed.colors.cyan);
    }
    screen.render();
}

function update() {
    if (gameOver) {
        draw();
        return;
    }

    // Physics
    velY += gravity;
    playerY += velY;
    playerX += velX;
    if (playerX < 0) playerX = 0;
    if (playerX >= width) playerX = width - 1;

    // Collision
    for (const p of platforms) {
        if (p.x <= playerX && playerX <= p.x+4 &&
            p.y <= playerY+1 && playerY+1 <= p.y+1 &&
            velY > 0) {
            velY = jump;
            process.stdout.write('\x07');
            if (p.type === 3) {
                score += 10;
                p.type = -1;
            } else if (p.type === 2) {
                p.type = -1;
            }
            break;
        }
    }
    platforms = platforms.filter(p => p.type !== -1);

    // Generate new platforms
    if (playerY < height/2) {
        const diff = height/2 - playerY;
        playerY += diff;
        for (const p of platforms) p.y += diff;
        while (platforms.length > 0 && platforms[platforms.length-1].y < height-2) {
            let x = Math.floor(Math.random() * (width-6)) + 1;
            let y = platforms[platforms.length-1].y - (Math.floor(Math.random()*4)+3);
            let r = Math.random()*100;
            let type = 0;
            if (r < 70) type = 0;
            else if (r < 85) type = 1;
            else if (r < 95) type = 2;
            else type = 3;
            platforms.push({x, y, type});
            score++;
        }
    }

    if (playerY > height) {
        gameOver = true;
        if (score > best) { best = score; saveRecord(best); }
    }
    draw();
    setTimeout(update, frameTime);
}

// Controls
screen.key(['left', 'a', 'A'], function() { velX = -3; });
screen.key(['right', 'd', 'D'], function() { velX = 3; });
screen.key(['space'], function() { velY = 2; });
screen.key(['r', 'R'], function() {
    if (gameOver) {
        playerY = height - 3;
        playerX = Math.floor(width/2);
        velX = 0; velY = 0;
        platforms = [];
        score = 0;
        gameOver = false;
        for (let i=0; i<8; i++) {
            let x = Math.floor(Math.random() * (width-6)) + 1;
            let y = height - 2 - i*4;
            platforms.push({x, y, type: 0});
        }
        draw();
    }
});
screen.key(['q', 'Q'], function() { process.exit(0); });

draw();
update();

screen.on('resize', function() {});

File: doodle.go
// doodle.go
package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"time"
	"github.com/nsf/termbox-go"
)

const recordFile = ".doodle_record.json"

type Record struct {
	Best int `json:"record"`
}

func loadRecord() int {
	f, err := os.Open(recordFile)
	if err != nil {
		return 0
	}
	defer f.Close()
	var r Record
	json.NewDecoder(f).Decode(&r)
	return r.Best
}

func saveRecord(best int) {
	r := Record{Best: best}
	f, _ := os.Create(recordFile)
	defer f.Close()
	json.NewEncoder(f).Encode(r)
}

func main() {
	speed := 50
	if len(os.Args) > 2 && os.Args[1] == "-s" {
		if s, err := strconv.Atoi(os.Args[2]); err == nil && s > 0 {
			speed = s
		}
	}
	err := termbox.Init()
	if err != nil {
		fmt.Println("termbox init failed:", err)
		return
	}
	defer termbox.Close()
	termbox.SetInputMode(termbox.InputEsc)
	w, h := termbox.Size()
	if h < 20 || w < 30 {
		fmt.Println("Terminal too small")
		return
	}
	rand.Seed(time.Now().UnixNano())

	playerY := float64(h - 3)
	playerX := float64(w / 2)
	velX, velY := 0.0, 0.0
	const gravity, jump = 0.3, -6.0
	type Platform struct{ x, y, typ int }
	platforms := []Platform{}
	score, best := 0, loadRecord()
	gameOver := false
	frame := time.Duration(speed) * time.Millisecond

	// initial platforms
	for i := 0; i < 8; i++ {
		x := rand.Intn(w-6) + 1
		y := h - 2 - i*4
		platforms = append(platforms, Platform{x, y, 0})
	}

	for {
		ev := termbox.PollEvent()
		if ev.Type == termbox.EventKey {
			if ev.Key == termbox.KeyEsc || ev.Ch == 'q' {
				return
			}
			if ev.Ch == 'r' {
				if gameOver {
					playerY = float64(h - 3)
					playerX = float64(w / 2)
					velX, velY = 0, 0
					platforms = []Platform{}
					score = 0
					gameOver = false
					for i := 0; i < 8; i++ {
						x := rand.Intn(w-6) + 1
						y := h - 2 - i*4
						platforms = append(platforms, Platform{x, y, 0})
					}
					continue
				}
			}
			if ev.Key == termbox.KeyArrowLeft || ev.Ch == 'a' {
				velX = -3
			} else if ev.Key == termbox.KeyArrowRight || ev.Ch == 'd' {
				velX = 3
			} else {
				velX *= 0.85
			}
			if ev.Ch == ' ' {
				velY = 2
			}
		}

		if gameOver {
			termbox.Clear(termbox.ColorDefault, termbox.ColorDefault)
			msg := fmt.Sprintf("💀 Game Over! Score: %d  Best: %d", score, best)
			tbprint(w/2-len(msg)/2, h/2-2, termbox.ColorRed, termbox.ColorDefault, msg)
			tbprint(w/2-10, h/2, termbox.ColorCyan, termbox.ColorDefault, "R - restart | Q - quit")
			termbox.Flush()
			continue
		}

		// Physics
		velY += gravity
		playerY += velY
		playerX += velX
		if playerX < 0 {
			playerX = 0
		}
		if playerX >= float64(w) {
			playerX = float64(w - 1)
		}

		// Collision
		for i := range platforms {
			p := &platforms[i]
			if float64(p.x) <= playerX && playerX <= float64(p.x+4) &&
				float64(p.y) <= playerY+1 && playerY+1 <= float64(p.y+1) &&
				velY > 0 {
				velY = jump
				fmt.Print("\a")
				if p.typ == 3 {
					score += 10
					p.typ = -1
				} else if p.typ == 2 {
					p.typ = -1
				}
				break
			}
		}
		// remove marked
		newPlats := []Platform{}
		for _, p := range platforms {
			if p.typ != -1 {
				newPlats = append(newPlats, p)
			}
		}
		platforms = newPlats

		// generate new platforms
		if playerY < float64(h/2) {
			diff := float64(h/2) - playerY
			playerY += diff
			for i := range platforms {
				platforms[i].y += int(diff)
			}
			for len(platforms) > 0 && platforms[len(platforms)-1].y < h-2 {
				x := rand.Intn(w-6) + 1
				y := platforms[len(platforms)-1].y - (rand.Intn(4) + 3)
				typ := rand.Intn(100)
				ptype := 0
				if typ < 70 {
					ptype = 0
				} else if typ < 85 {
					ptype = 1
				} else if typ < 95 {
					ptype = 2
				} else {
					ptype = 3
				}
				platforms = append(platforms, Platform{x, y, ptype})
				score++
			}
		}

		if playerY > float64(h) {
			gameOver = true
			if score > best {
				best = score
				saveRecord(best)
			}
		}

		// Draw
		termbox.Clear(termbox.ColorDefault, termbox.ColorDefault)
		// player
		termbox.SetCell(int(playerX), int(playerY), '@', termbox.ColorYellow|termbox.AttrBold, termbox.ColorDefault)
		// platforms
		for _, p := range platforms {
			var fg termbox.Attribute
			switch p.typ {
			case 1:
				fg = termbox.ColorGreen | termbox.AttrBold
			case 2, 3:
				fg = termbox.ColorRed
			default:
				fg = termbox.ColorGreen
			}
			for i := 0; i < 5; i++ {
				if p.x+i >= 0 && p.x+i < w {
					termbox.SetCell(p.x+i, p.y, '#', fg, termbox.ColorDefault)
				}
			}
		}
		// score
		tbprint(2, 0, termbox.ColorCyan, termbox.ColorDefault, fmt.Sprintf("Score: %d", score))
		tbprint(w/2-4, 0, termbox.ColorCyan, termbox.ColorDefault, fmt.Sprintf("Best: %d", best))
		termbox.Flush()
		time.Sleep(frame)
	}
}

func tbprint(x, y int, fg, bg termbox.Attribute, msg string) {
	for _, ch := range msg {
		termbox.SetCell(x, y, ch, fg, bg)
		x++
	}
}
